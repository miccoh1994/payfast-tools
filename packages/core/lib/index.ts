export * from './payfast-types';
export * from './generate-signature';