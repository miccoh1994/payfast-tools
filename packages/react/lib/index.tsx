export * from './main';
export * from './use-payfast';