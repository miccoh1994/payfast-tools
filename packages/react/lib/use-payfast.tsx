import {
  DetailedHTMLProps,
  InputHTMLAttributes,
  ReactNode, 
  useMemo,
} from 'react';
import { generateSignature } from '@miccoh/payfast-core';
import type { CustomerDetails, MerchantDetails, TransactionDetails, TransactionOptions } from '@miccoh/payfast-core';

  export interface ReactPayfastProps {
    sandbox?: boolean;
    merchant: MerchantDetails;
    customer: CustomerDetails;
    transaction: TransactionDetails;
    options?: TransactionOptions;
    children?: ReactNode;
  }
  
  type FieldValue = string | number | readonly string[] | undefined;
  export function usePayfast({
    merchant,
    customer,
    transaction,
    options,
  }: ReactPayfastProps) {
    const mergedOpts = {
      ...merchant,
      ...customer,
      ...transaction,
      ...options,
    };

    const signature = useMemo(() => {
      return generateSignature({
        ...merchant,
        ...customer,
        ...transaction,
        ...options,
      });
    },  [customer, merchant, options, transaction])


    const fields: DetailedHTMLProps<
      InputHTMLAttributes<HTMLInputElement>,
      HTMLInputElement
    >[] = Object.entries(mergedOpts).map(([key, value]) => {
      return {
        type: 'hidden',
        name: key,
        value: value as FieldValue,
      };
    });
    return {
      signature,
      fields: [...fields],
    };
  }
  