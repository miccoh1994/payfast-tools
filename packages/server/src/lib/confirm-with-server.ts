import type { ISignatureInput } from '@miccoh/payfast-core';
import axios from 'axios';
import { getParameterString } from '@miccoh/payfast-core';

export async function confirmWithServer(
  host: string,
  data: ISignatureInput,
  passphrase?: string
) {
  const parameterStringWithSig = getParameterString(data, passphrase);

  const { data: result } = await axios.post(
    `https://${host}/eng/query/validate`,
    parameterStringWithSig
  );
  if (result === 'VALID') {
    return true;
  }
  console.warn(
    `[payfast]: failed to validate with server, host: ${host} \n error: ${result}`
  );
  return false;
}
