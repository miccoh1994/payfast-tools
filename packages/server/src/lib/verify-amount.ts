export function verifyAmount(
  mAmount: number | string,
  pfAmountGross: number | string
) {
  const validAmount =
    Math.abs(parseFloat(String(mAmount)) - parseFloat(String(pfAmountGross))) <=
    0.01;
  if (validAmount === true) {
    return validAmount;
  }
  console.warn('[payfast]: failed to verify amount');
  return validAmount;
}
