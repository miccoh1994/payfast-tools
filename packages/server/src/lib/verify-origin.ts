import { lookup } from 'dns';
const VALID_ORIGINS = [
  'www.payfast.co.za',
  'sandbox.payfast.co.za',
  'w1w.payfast.co.za',
  'w2w.payfast.co.za',
];

export function dnsLookup(origin: string): Promise<string[]> {
  return new Promise((resolve, reject) => {
    lookup(origin, { all: true }, (err, address) => {
      if (err) {
        reject(err);
        return;
      }
      resolve(address.map((add) => add.address));
    });
  });
}
/**
 * 
 * @param xForwardedFor req.headers['x-forwarded-for'] || req.socket.remoteAddress;
 * @returns 
 */
export async function verifyOrigin(xForwardedFor: string): Promise<boolean> {
  if (process.env.NODE_ENV === 'development') {
    xForwardedFor = (xForwardedFor as string).split(',')[0];
  }
  try {
    const validIps = await Promise.all(
      VALID_ORIGINS.map((origin) => {
        return dnsLookup(origin);
      })
    );
    if (validIps.flat().includes(xForwardedFor as string)) {
      return true;
    }
    console.warn('[payfast]: failed to validate origin');
    return false;
  } catch (e) {
    console.log(e);
    return false;
  }
}
