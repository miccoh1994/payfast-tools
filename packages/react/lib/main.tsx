import { ReactPayfastProps, usePayfast } from './use-payfast'

export const ReactPayfast = ({
  sandbox,
  children,
  ...props
}: ReactPayfastProps) => {
  const { fields } = usePayfast(props);
  const actionUrl = sandbox
    ? 'https://sandbox.payfast.co.za/eng/process'
    : 'https://www.payfast.co.za/eng/process';
  return (
    <form name="react-payfast" action={actionUrl} method="post">
      {fields.map((field) => (
        <input
          {...field}
          key={field.name}
          role="none"
          data-testid={field.name}
        />
      ))}
      {children}
    </form>
  );
};

export default ReactPayfast;
