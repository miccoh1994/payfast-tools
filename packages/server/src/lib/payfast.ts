import { confirmWithServer } from './confirm-with-server';
import { verifyAmount } from './verify-amount';
import { verifyOrigin } from './verify-origin';
import { verifySignature } from './verify-signature';
import type { ISignatureInput } from '@miccoh/payfast-core';

export interface IValidatePaymentOptions {
  yourAmount: number;
  passphrase?: string;
  sandbox?: boolean;
}
export async function validatePayment(
  {
    xForwardedForHeader,
    signature,
    payload,
  }: {
    xForwardedForHeader: string;
    signature?: string;
    payload: ISignatureInput;
  },
  { yourAmount, passphrase, sandbox }: IValidatePaymentOptions
) {
  try {
    const validSignature = verifySignature(signature, payload, passphrase);
    const validIp = await verifyOrigin(xForwardedForHeader);
    const validData = verifyAmount(yourAmount, payload.amount_gross);
    const host = sandbox === true ? 'sandbox.payfast.co.za' : 'payfast.co.za';
    const serverConfirmed = await confirmWithServer(host, payload, passphrase);

    if (
      validSignature === true &&
      validIp === true &&
      validData === true &&
      serverConfirmed === true
    ) {
      return true;
    }
    console.error('[payfast]: failed to validate payment');
    return false;
  } catch (e) {
    console.error(e);
    return false;
  }
}
