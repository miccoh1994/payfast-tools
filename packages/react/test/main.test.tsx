/* eslint-disable camelcase */
import { test, expect } from "vitest";
import renderer from 'react-test-renderer'
import {ReactPayfast} from '../lib';
import { nanoid } from "nanoid";
function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}
const buildOptions = (m_payment_id: string, notify_url: string) => ({
  merchant: {
    merchant_id: '10000100',
    merchant_key: '46f0cd694581a',
    return_url:  '/success',
    cancel_url: '/success',
    notify_url,
  },
  customer: {
    name_first: 'Test',
    name_last: 'Test',
    email_address: 'test@test.com',
  },
  transaction: {
    m_payment_id,
    amount: 100,
    item_name: 'cookies',
    item_description: 'a bag of cookies',
  },
});
const options = buildOptions(nanoid(), 'url')
test('MyEpicLibraryComponent renders succesfully', () => {
  const component = renderer.create(
    <ReactPayfast {...options}/>,
  )
  const tree = toJson(component)
  expect(tree).toMatchSnapshot()
})