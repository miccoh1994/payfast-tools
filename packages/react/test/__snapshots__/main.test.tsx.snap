// Vitest Snapshot v1

exports[`MyEpicLibraryComponent renders succesfully 1`] = `
<form
  action="https://www.payfast.co.za/eng/process"
  method="post"
  name="react-payfast"
>
  <input
    data-testid="merchant_id"
    name="merchant_id"
    role="none"
    type="hidden"
    value="10000100"
  />
  <input
    data-testid="merchant_key"
    name="merchant_key"
    role="none"
    type="hidden"
    value="46f0cd694581a"
  />
  <input
    data-testid="return_url"
    name="return_url"
    role="none"
    type="hidden"
    value="/success"
  />
  <input
    data-testid="cancel_url"
    name="cancel_url"
    role="none"
    type="hidden"
    value="/success"
  />
  <input
    data-testid="notify_url"
    name="notify_url"
    role="none"
    type="hidden"
    value="url"
  />
  <input
    data-testid="name_first"
    name="name_first"
    role="none"
    type="hidden"
    value="Test"
  />
  <input
    data-testid="name_last"
    name="name_last"
    role="none"
    type="hidden"
    value="Test"
  />
  <input
    data-testid="email_address"
    name="email_address"
    role="none"
    type="hidden"
    value="test@test.com"
  />
  <input
    data-testid="m_payment_id"
    name="m_payment_id"
    role="none"
    type="hidden"
    value="uJ-su7sNIpXN3Q_0v1Z8w"
  />
  <input
    data-testid="amount"
    name="amount"
    role="none"
    type="hidden"
    value={100}
  />
  <input
    data-testid="item_name"
    name="item_name"
    role="none"
    type="hidden"
    value="cookies"
  />
  <input
    data-testid="item_description"
    name="item_description"
    role="none"
    type="hidden"
    value="a bag of cookies"
  />
</form>
`;
